src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -596,106 +924,289 @@ namespace nbl::ext::imgui
-
-					assert(vBinding.memory->isCurrentlyMapped());
+					const ImDrawList* commandList = drawData->CmdLists[i];
+					params.bytesToFill[MDI::EBC_DRAW_INDIRECT_STRUCTURES] += commandList->CmdBuffer.Size * sizeof(VkDrawIndexedIndirectCommand);
+					params.bytesToFill[MDI::EBC_ELEMENT_STRUCTURES] += commandList->CmdBuffer.Size * sizeof(PerObjectData);
 				}
+				params.bytesToFill[MDI::EBC_VERTEX_BUFFERS] = drawData->TotalVtxCount * sizeof(ImDrawVert);
+				params.bytesToFill[MDI::EBC_INDEX_BUFFERS] = drawData->TotalIdxCount * sizeof(ImDrawIdx);
+
+				// calculate upper bound byte size limit for mdi buffer
+				params.totalByteSizeRequest = std::reduce(std::begin(params.bytesToFill), std::end(params.bytesToFill));
+				params.drawCount = params.bytesToFill[MDI::EBC_DRAW_INDIRECT_STRUCTURES] / sizeof(VkDrawIndexedIndirectCommand);
+
+				return std::move(params);
+			}();
 
+			std::array<bool, MDI_COMPONENT_COUNT> mdiBytesFilled;
+			std::fill(mdiBytesFilled.data(), mdiBytesFilled.data() + MDI_COMPONENT_COUNT, false);
+			std::array<MDI_SIZE_TYPE, MDI_COMPONENT_COUNT> mdiOffsets;
+			std::fill(mdiOffsets.data(), mdiOffsets.data() + MDI_COMPONENT_COUNT, INVALID_ADDRESS);
+
+			auto streamingBuffer = m_mdi.buffer;
+			{
+				auto binding = streamingBuffer->getBuffer()->getBoundMemory();
+				assert(binding.memory->isCurrentlyMapped());
+
+				auto* const mdiData = reinterpret_cast<uint8_t*>(streamingBuffer->getBufferPointer());
+
+				struct
 				{
-					if (!iBinding.memory->map({ 0ull, iBinding.memory->getAllocationSize() }, IDeviceMemoryAllocation::EMCAF_READ))
-						logger->log("Could not map device memory for index buffer data!", system::ILogger::ELL_WARNING);
-				
-					assert(iBinding.memory->isCurrentlyMapped());
-				}
+					MDI_SIZE_TYPE offset = INVALID_ADDRESS,
+					multiAllocationSize = {};
+				} requestState;
 
-				auto* vertex_ptr = static_cast<ImDrawVert*>(vBinding.memory->getMappedPointer());
-				auto* index_ptr = static_cast<ImDrawIdx*>(iBinding.memory->getMappedPointer());
+				const auto start = std::chrono::steady_clock::now();
 
-				for (int n = 0; n < drawData->CmdListsCount; n++)
+				//! we must upload entire MDI buffer data to our streaming buffer, but we cannot guarantee allocation can be done in single request
+				for (MDI_SIZE_TYPE uploadedSize = 0ull; uploadedSize < mdiParams.totalByteSizeRequest;)
 				{
-					const ImDrawList* cmd = drawData->CmdLists[n];
-					::memcpy(vertex_ptr, cmd->VtxBuffer.Data, cmd->VtxBuffer.Size * sizeof(ImDrawVert));
-					::memcpy(index_ptr, cmd->IdxBuffer.Data, cmd->IdxBuffer.Size * sizeof(ImDrawIdx));
-					vertex_ptr += cmd->VtxBuffer.Size;
-					index_ptr += cmd->IdxBuffer.Size;
-				}
+					auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start);
+
+					if (elapsed.count() >= 1u)
+					{
+						streamingBuffer->cull_frees();
+						return false;
+					}
```

Comment:
1. do not check for the wait point passing before doing the first iteration, this way you can end up with nasty flickers (all our utils to at least one try before timing out)
2. the elapsed check and `cull_frees` can move to the failure case (if elapsed, return false, if not then continue)
3. I don't think you need to call `cull_frees`